record_batch.go

```diff
@@ -0,0 +1,204 @@
+package kafka
+
+import (
+	"bufio"
+	"bytes"
+	"hash/crc32"
+	"time"
+)
+
+type record struct {
+	attributes int8
+	baseTime   time.Time
+	offset     int64
+	msg        Message
+}
+
+func (r record) size() int32 {
+	timestampDelta := r.msg.Time.Sub(r.baseTime)
+	offsetDelta := int64(r.offset)
+
+	var res int
+	res += 1 + // attributes
+		varIntLen(int64(milliseconds(timestampDelta))) +
+		varIntLen(offsetDelta) +
+		varIntLen(int64(len(r.msg.Key))) +
+		len(r.msg.Key) +
+		varIntLen(int64(len(r.msg.Value))) +
+		len(r.msg.Value) +
+		varIntLen(int64(len(r.msg.Headers)))
+	for _, h := range r.msg.Headers {
+		res += varIntLen(int64(len([]byte(h.Key)))) +
+			len([]byte(h.Key)) +
+			varIntLen(int64(len(h.Value))) +
+			len(h.Value)
+	}
+	return int32(res)
+}
+
+func (r record) writeTo(w *bufio.Writer) {
+	timestampDelta := r.msg.Time.Sub(r.baseTime)
+	offsetDelta := int64(r.offset)
+
+	writeVarInt(w, int64(r.size()))
+
+	writeInt8(w, r.attributes)
+	writeVarInt(w, int64(milliseconds(timestampDelta)))
+	writeVarInt(w, offsetDelta)
+
+	writeVarInt(w, int64(len(r.msg.Key)))
+	w.Write(r.msg.Key)
+	writeVarInt(w, int64(len(r.msg.Value)))
+	w.Write(r.msg.Value)
+	writeVarInt(w, int64(len(r.msg.Headers)))
+
+	for _, h := range r.msg.Headers {
+		writeVarInt(w, int64(len(h.Key)))
+		w.Write([]byte(h.Key))
+		writeVarInt(w, int64(len(h.Value)))
+		w.Write(h.Value)
+	}
+}
+
+type recordBatch struct {
+	codec      CompressionCodec
+	attributes int16
+	msgs       []Message
+	compressed []byte
+}
+
+func (r *recordBatch) Init() (err error) {
+	if r.codec != nil {
+		r.attributes = int16(r.codec.Code()) | r.attributes
+		recordBuf := &bytes.Buffer{}
+		recordBuf.Grow(int(recordBatchSize(r.msgs...)))
+		compressedWriter := bufio.NewWriter(recordBuf)
+		for i, msg := range r.msgs {
+			record{
+				attributes: 0,
+				baseTime:   r.msgs[0].Time,
+				offset:     int64(i),
+				msg:        msg,
+			}.writeTo(compressedWriter)
+		}
+		compressedWriter.Flush()
+
+		r.compressed, err = r.codec.Encode(recordBuf.Bytes())
+		if err != nil {
+			return
+		}
+		r.attributes = int16(r.codec.Code())
+	}
+	return nil
+}
+
+func (r recordBatch) size() (size int32) {
```

Comment:
```suggestion
func (r *recordBatch) size() (size int32) {
```
It looks like we're using these values through pointers (https://github.com/segmentio/kafka-go/pull/306), how about keeping things consistent and pass the value by pointer everywhere?

Reply:
I'm using non pointer receivers when I want to communicate that this is a const method - method that doesn't modify the structure. If you insist I can always use pointer receivers.

Reply:
I assumed that this was the intention, however in practice I found that mixing pointer and value receiver on a single type tends to bring more confusion to the readers. I feel like a type should either be immutable, in which case all it’s method should have value receivers, or mutable and have all its methods have pointer receivers. 

Consistency is a powerful way of guiding people’s intuition to follow the intent of the original author, and helps them be more successful with the tools we build.

Reply:
I see your point. Pushed the fix.